### DISCUSSION TITLE
getting results from IBM DB2 stored procedure

### DISCUSSION THREAD

**Discussion by DeltaFox0018** [mkleehammer/pyodbc#1384] [Q&A]
Please first make sure you have looked at:

* Documentation: https://github.com/mkleehammer/pyodbc/wiki
* Other issues

### Environment

To diagnose, we usually need to know the following, including version numbers.  On Windows, be
sure to specify 32-bit Python or 64-bit:

- Python: 3.12.6
- pyodbc: 5.10
- OS: MacOS 15.0.1 (M1 Pro)
- DB: IBM DB2 (IBM Power 9)
- driver: IBM i Access ODBC Driver

### Issue

I have a procedure with 3 parameters, 2 input and 1 output.

From DBeaver if I call the procedure as follows `CALL my_procedure('user','psw', ?)`

I have an output of the following type
<img width="150" alt="image" src="https://github.com/user-attachments/assets/cbacad2a-7b91-43da-a2c0-247dde947749">

Through pyodbc I configured the connection to the DB and ran the same call

`    
 cnxn = pyodbc.connect("DRIVER=IBM i Access ODBC Driver;SYSTEM=MYIP;UID=USER;PWD=PSW;")

    cursor = cnxn.cursor()

    # Parameters for the stored procedure
    params = ('user', "psw", 0)

    cursor.execute("{CALL my_procedure (?, ?, ?)}", params)
    print(cursor.fetchall())
`

Running the print either you fetchall, fetchone or fetchval or the following error

`
Traceback (most recent call last):
  File "/Users/alessandrovolpato/Desktop/rapportini-backend/main.py", line 43, in <module>
    print(cursor.fetchone())
          ^^^^^^^^^^^^^^^^^
pyodbc.ProgrammingError: No results.  Previous SQL was not a query.
`

Do you know how I can retrieve the output of the procedure?

**Comment by gordthompson** [comment]
You might try this:

```python
cursor.execute(sql)
try:
    rows = cursor.fetchall()
except pyodbc.ProgrammingError:
    cursor.nextset()
    rows = cursor.fetchall()
print(rows)
```

**Comment by DeltaFox0018** [comment]
I have tried but the error is the same

**Comment by gordthompson** [comment]
A more robust approach would be

```python
cursor.execute(sql)
rows = []
fetching = True
while fetching:
    try:
        rows = cursor.fetchall()
        fetching = False
    except pyodbc.ProgrammingError:
        fetching = cursor.nextset()
print(rows)
```

**Comment by kadler** [comment]
pyodbc doesn't support output parameters. There's a workaround documented [here](https://github.com/mkleehammer/pyodbc/wiki/Calling-Stored-Procedures)

**Reply by kadler** [reply]
FYI you would need to do something a little different on Db2 since you cannot just declare a variable and you also cannot execute multiple statements as a script:

```python
# Sample procedure
conn.execute("""
create or replace procedure kadler.test_outparm(x out varchar(10))
begin
    set x = 'hello';
end
""")

# Unfortunately, QTEMP not allowed for variables
conn.execute("create variable kadler.test_outparm_out varchar(10)");

# Call the procedure, output goes to the variable created earlier
conn.execute("call kadler.test_outparm(kadler.test_outparm_out)")

# Get the output
output = conn.execute("values(kadler.test_outparm_out)").fetchval()
```

It's a bit fiddly, but this is the best workaround that I know of. Another alternative would be to replace/wrap the procedure with an SQL scalar function, but that only works if you have a single output parameter and don't need to return any cursors.

**Reply by kadler** [reply]
An example of the scalar function wrapper for this case might be:

```sql
create function my_function(user varchar(10), pass varchar(128))
returns varchar(10) modifies sql data
begin
   declare output varchar(10);
   call my_procedure(user, pass, output);
   return output;
end
```

Then you'd call it like so:

```python
output = conn.execute("values(my_function(?, ?))", ('user', 'psw')).fetchval()
```

**Reply by muggel042** [reply]
Hello @kadler 

I am delighted to find this thread. I've already spent a lot of time to understand stored procedures, API calls and pyodbc. It feels like I have worked it out and only a tiny bit is wrong.

Here is what I got and what hopefully may help others too (if corrected)

```
#create or replace stored procedure for API call 
sql = "CREATE OR REPLACE PROCEDURE TESTSQL.RETRIEVE_CERT_USAGE_INFO (
                OUT RECEIVER_VARIABLE CHAR(32750) FOR BIT DATA,
                IN LENGTH_OF_RECEIVER_VARIABLE INTEGER,
                IN FORMAT_NAME CHAR(8),
                IN APPLICATION_SELECTION_CRITERIA CHAR(4) FOR BIT DATA,
                IN ERROR_CODE CHAR(4)
             )
             LANGUAGE CL
             DYNAMIC RESULT SETS 1
             NOT DETERMINISTIC
             READS SQL DATA
             CALLED ON NULL INPUT
             EXTERNAL NAME 'QICSS/QYCDCUSG(QycdRetrieveCertUsageInfo)'
             PARAMETER STYLE GENERAL"

db.execute(sql)
#create output variable
db.execute("CREATE OR REPLACE VARIABLE TESTSQL.DCM_OUT CHAR(32750)")

#execute stored procedure with all parameters
sql = "{CALL TESTSQL.RETRIEVE_CERT_USAGE_INFO(?,?,?,?,?)}"
params = ('TESTSQL.DCM_OUT', 32750 ,'RCUI0200', bin(0), '')
db.execute(sql,params)

#fetch variable content
data = db.execute("VALUES(TESTSQL.DCM_OUT)").fetchval()

print(data)
```

But this currently fails with:
('22018', '[22018] [IBM][System i Access ODBC Driver]Error in assignment. (30019) (SQLExecDirectW)') 

And yes ... I try to get information about if an application has assigned a ssl certificate in DCM which is still valid. There is an IBM Ideas up on this topic, but only god knows when this will be implemented and for what releases.
https://ideas.ibm.com/ideas/IBMI-I-4104

**Reply by kadler** [reply]
You are passing "TESTSQL.DCM_OUT" in to a parameter marker. That will attempt to pass the literal value "TESTSQL.DCM_OUT" to that store procedure, **not** a reference to the variable named TESTSQL.DCM_OUT.

I think you'd need to change your SQL like so: `sql = "{CALL TESTSQL.RETRIEVE_CERT_USAGE_INFO(TESTSQL.DCM_OUT,?,?,?,?)}"`

**Reply by muggel042** [reply]
Thank you. It's working a little bit better, but the fetchval() function gives always "none" as a result.

I think I'll give up on this and wait for an official DB2 function or table on this.

**Reply by kadler** [reply]
A variable has to be initialized in the job before it can be used. If you do a `SET TESTSQL.DCM_OUT = ''` before calling 
TESTSQL.RETRIEVE_CERT_USAGE_INFO, that will activate the variable in your job and then when you query it, it'll have a value.

**Reply by kadler** [reply]
I would also change APPLICATION_SELECTION_CRITERIA and ERROR_CODE to INTEGER and pass 0, eg.

```python
#create or replace stored procedure for API call 
sql = """CREATE OR REPLACE PROCEDURE TESTSQL.RETRIEVE_CERT_USAGE_INFO (
                OUT RECEIVER_VARIABLE CHAR(32750) FOR BIT DATA,
                IN LENGTH_OF_RECEIVER_VARIABLE INTEGER,
                IN FORMAT_NAME CHAR(8),
                IN APPLICATION_SELECTION_CRITERIA INTEGER,
                IN ERROR_CODE INTEGER
             )
             LANGUAGE CL
             DYNAMIC RESULT SETS 1
             NOT DETERMINISTIC
             READS SQL DATA
             CALLED ON NULL INPUT
             EXTERNAL NAME 'QICSS/QYCDCUSG(QycdRetrieveCertUsageInfo)'
             PARAMETER STYLE GENERAL"""

db.execute(sql)
#create output variable
db.execute("CREATE OR REPLACE VARIABLE TESTSQL.DCM_OUT CHAR(32750)")
db.execute("SET TESTSQL.DCM_OUT = ''")

#execute stored procedure with all parameters, except output parameter global variable
sql = "{CALL TESTSQL.RETRIEVE_CERT_USAGE_INFO(TESTSQL.DCM_OUT,?,?,?,?)}"
params = (32750 ,'RCUI0200', 0, 0)
db.execute(sql,params)

#fetch variable content
data = db.execute("VALUES(TESTSQL.DCM_OUT)").fetchval()

print(data)
```

**Reply by muggel042** [reply]
Hi

I already knew you're the best, but please let me say it again: You're the best!

With your modifications I receive output, which I can now process.

Many thanks for all the help.
